lnbits/core/services/payments.py

```diff
@@ -195,21 +196,54 @@ def service_fee(amount_msat: int, internal: bool = False) -> int:
         return 0
 
 
-async def update_wallet_balance(wallet_id: str, amount: int):
-    async with db.connect() as conn:
-        payment = await create_invoice(
-            wallet_id=wallet_id,
-            amount=amount,
-            memo="Admin top up",
-            internal=True,
-            conn=conn,
-        )
-        payment.status = PaymentState.SUCCESS
-        await update_payment(payment, conn=conn)
-        # notify receiver asynchronously
-        from lnbits.tasks import internal_invoice_queue
+async def update_wallet_balance(
+    wallet: Wallet,
+    amount: int,
+    conn: Optional[Connection] = None,
+):
 
-        await internal_invoice_queue.put(payment.checking_id)
+    if amount == 0:
+        raise ValueError("Amount cannot be 0.")
+    if (
+        settings.lnbits_wallet_limit_max_balance > 0
+        and wallet.balance + amount > settings.lnbits_wallet_limit_max_balance
+    ):
+        raise ValueError("Balance change failed, amount exceeds maximum balance.")
+
+    # negative topup
+    if amount < 0:
+        if wallet.balance_msat + amount < 0:
+            raise ValueError("Balance change failed, can not go into negative balance.")
+
+    async with db.reuse_conn(conn) if conn else db.connect() as conn:
+        if amount > 0:
+            payment = await create_invoice(
+                wallet_id=wallet.id,
+                amount=amount,
+                memo="Admin top up" if amount < 0 else "Admin withdrawal",
+                internal=True,
+                conn=conn,
+            )
+            payment.status = PaymentState.SUCCESS
+            await update_payment(payment, conn=conn)
+            # notify receiver asynchronously
+            from lnbits.tasks import internal_invoice_queue
```

Comment:
- we should avoid import in the middle of the code, can the queue be an optional input param instead?

Reply:
can you look at that in another PR? it was like that before and its related to circular imports

Reply:
ok